src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+      lderr(m_cct) << "error releasing exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+  } else {
+    m_image_ctx->owner_lock.put_read();
+  }
+
+  int r = trash_move(m_src_io_ctx, RBD_TRASH_IMAGE_SOURCE_MIGRATION,
+                     m_image_ctx->name, 0);
+  if (r < 0) {
+    lderr(m_cct) << "failed moving image to trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::v2_relink_image() {
+  int r = trash_restore(m_src_io_ctx, m_image_ctx->id, m_image_ctx->name);
+  if (r < 0) {
+    lderr(m_cct) << "failed restoring image from trash: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  return 0;
+}
+
+template <typename I>
+int Migrate<I>::create_new_image() {
+  ldout(m_cct, 10) << dendl;
+
+  uint64_t size;
+  {
+    RWLock::RLocker snap_locker(m_image_ctx->snap_lock);
+    size = m_image_ctx->size;
+  }
+
+  ThreadPool *thread_pool;
+  ContextWQ *op_work_queue;
+  ImageCtx::get_thread_pool_instance(m_cct, &thread_pool, &op_work_queue);
+
+  C_SaferCond on_create;
+  auto *req = image::CreateRequest<I>::create(
+      m_dst_io_ctx, m_dst_image_name, m_dst_image_id, size, m_image_options, "",
+      "", false, op_work_queue, &on_create);
+  req->send();
+  int r = on_create.wait();
+  if (r < 0) {
+    lderr(m_cct) << "header creation failed: " << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  auto dst_image_ctx = new librbd::ImageCtx(m_dst_image_name, m_dst_image_id,
+                                            nullptr, m_dst_io_ctx, false);
+  dst_image_ctx->ignore_migrating = true; // XXXMG: hackish
+  r = dst_image_ctx->state->open(false);
+  if (r < 0) {
+    lderr(m_cct) << "failed to open newly created header: " << cpp_strerror(r)
+                 << dendl;
+    return r;
+  }
+
+  dst_image_ctx->owner_lock.get_read();
+  if (dst_image_ctx->exclusive_lock != nullptr &&
+      !dst_image_ctx->exclusive_lock->is_lock_owner()) {
+    C_SaferCond ctx;
+    dst_image_ctx->exclusive_lock->acquire_lock(&ctx);
+    dst_image_ctx->owner_lock.put_read();
+    r = ctx.wait();
+    if (r < 0) {
+      lderr(m_cct) << "error acquiring exclusive lock: " << cpp_strerror(r)
+                   << dendl;
+      return r;
+    }
+  } else {
+    dst_image_ctx->owner_lock.put_read();
+  }
+
+  SnapSeqs snap_seqs;
+
+  C_SaferCond on_snapshot_copy;
+  auto snapshot_copy_req = librbd::deep_copy::SnapshotCopyRequest<I>::create(
+      m_image_ctx, dst_image_ctx, m_image_ctx->op_work_queue, &snap_seqs,
+      &on_snapshot_copy);
+  snapshot_copy_req->send();
+  r = on_snapshot_copy.wait();
+  if (r < 0) {
+    lderr(m_cct) << "failed to copy snapshots: " << cpp_strerror(r) << dendl;
+    dst_image_ctx->state->close();
+    return r;
+  }
+
+  C_SaferCond on_set_head;
+  librbd::ParentSpec parent_spec = {};
+  uint64_t parent_overlap = 0;
+  auto set_head_req = librbd::deep_copy::SetHeadRequest<I>::create(
```

Comment:
Is this necessary? The image was just create as a standalone to (hopefully) the correct size.

Reply:
@dillaman  I forgot to write about this yesterday.
After `deep_copy::SnapshotCopyRequest`, the value of `dst_image_ctx->size` is set to the size of the latest snapshot. So I need to update it by calling `SetHeadRequest` (in case the image was resized after the last snapshot had been created).

Reply:
I wonder if we should just move the ```SetHeadRequest``` into the ```SnapshotCopyRequest``` state machine so it can be removed here and in ```DeepCopyRequest```

Reply:
I will try. Thanks

Reply:
@dillaman Looking at it now. In `DeepCopyRequest` we call `SetHeadRequest` only if snap_id_end is head. Right now `SnapshotCopyRequest` does not have information about snap_id_end. I could add this a parameter, but actually, I wander if it works correctly now if one wants to deep copy only a subset of snapshots? `SnapshotCopyRequest` ignores `snap_id_start/end` and creates all image snapshots. It looks to me it should create only `[snap_id_start, snap_id_end]` subset? Or at least up to `snap_id_end`?

Reply:
I would agree that it should be up to (and including) ```snap_id_end``` -- not sure if it ever makes sense to specify a non-zero ```snap_id_start``` (and maybe it should just be removed).

Reply:
... nevermind about my ```snap_id_start``` comment, I forgot about the incremental image sync case for rbd-mirror.